agent/connectionuptime/service.go

```diff
@@ -0,0 +1,128 @@
+// pmm-agent
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connectionuptime
+
+import (
+	"math/big"
+	"sync"
+	"time"
+)
+
+// Service calculates connection uptime between agent and server
+type Service struct {
+	mx sync.Mutex
+
+	bits big.Int
+
+	windowPeriodSeconds int64
+	indexLastStatus     int64
+	startTime           time.Time
+	lastStatusTimestamp time.Time
+}
+
+// NewService creates new instance of Service
+func NewService(windowPeriod time.Duration) *Service {
+	return &Service{
+		windowPeriodSeconds: int64(windowPeriod.Seconds()),
+	}
+}
+
+// RegisterConnectionStatus adds new connection status
+func (s *Service) RegisterConnectionStatus(timestamp time.Time, connected bool) {
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	s.registerConnectionStatus(timestamp, connected)
+}
+
+func (s *Service) registerConnectionStatus(timestamp time.Time, connected bool) {
+	if s.startTime.IsZero() {
+		s.startTime = timestamp
+		s.lastStatusTimestamp = timestamp
+		s.bits.SetBit(&s.bits, 0, getBit(connected))
+		s.indexLastStatus = 0
+
+		return
+	}
+
+	secondsFromLastEvent := timestamp.Unix() - s.lastStatusTimestamp.Unix()
+	for i := s.indexLastStatus + 1; i < (s.indexLastStatus + secondsFromLastEvent); i++ {
+		// set the same status to elements of previous connection status
+		s.bits.SetBit(&s.bits, int(i%s.windowPeriodSeconds), s.bits.Bit(int(s.indexLastStatus)))
+	}
+
+	s.indexLastStatus = (s.indexLastStatus + secondsFromLastEvent) % s.windowPeriodSeconds
+	s.bits.SetBit(&s.bits, int(s.indexLastStatus), getBit(connected))
```

Comment:
How can we handle that correctly if `SetBit()` method accepts only int?


In theory it's possible that s.indexLastStatus will be bigger then maximum of int32, but secondsFromLastEvent is calculated as difference between two Unix timestamps and practically I'm not sure that it will exceed 2^31 value.

Reply:
I just wanted to say that `s.indexLastStatus` can't exceed the size of bitset which is used for storing connections statuses (`s.windowPeriodSeconds`)